### PR TITLE
RPE-266: migration

### PR DESCRIPTION
**Pull request by doncem**
### JIRA link (if applicable) ###

[db migrations](https://tools.hmcts.net/jira/browse/RPE-266)

### Change description ###

gradle flyway plugin to run tasks against db

only a plugin so to separate jenkins pipeline. future task: figure out how to not disable flyway on spring boot app but disable migration. instead do up-to-date check

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -4,12 +4,19 @@ plugins {
...
   id 'org.springframework.boot' version '1.5.9.RELEASE'
   id 'org.owasp.dependencycheck' version '3.1.0'
   id 'com.github.ben-manes.versions' version '0.17.0'
   id 'org.sonarqube' version '2.6.1'
 }
 
+flyway {
+  url = System.getenv('FLYWAY_URL')
+  user = System.getenv('FLYWAY_USER')
+  password = System.getenv('FLYWAY_PASSWORD')
```
Comment (rkondratowicz) [inline build.gradle]: worth adding `DB` for clarity imho. ie: `FLYWAY_DB_USER` etc

**Comment by rkondratowicz** [comment]
where's the migration? :P

File: build.gradle (hunk 1 above)
Comment (doncem) [reply build.gradle]: these are flyway's environment variables in use. if we remove this block, we simply do this:

```bash
$ FLYWAY_URL=jdbc:h2:mem:temp.db ./gradlew flywayInfo

> Task :flywayInfo 
Schema version: << Empty Schema >>
+-----------+---------+-------------+------+--------------+---------+
| Category  | Version | Description | Type | Installed On | State   |
+-----------+---------+-------------+------+--------------+---------+
| Versioned | 1       | Init        | SQL  |              | Pending |
+-----------+---------+-------------+------+--------------+---------+

```

File: src/main/resources/db/migration/V1__Init.sql [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  sent_to_print TIMESTAMP NULL,
+  printed TIMESTAMP NULL
+);
+
+CREATE TABLE letter_metadata (
+  id BIGSERIAL NOT NULL PRIMARY KEY,
+  letter_id UUID NOT NULL,
+  metadata VARCHAR(256) NOT NULL,
+  metadata_value VARCHAR(256) NOT NULL
+);
```
Comment (rkondratowicz) [inline src/main/resources/db/migration/V1__Init.sql]: we can simply store additional data in json column

File: src/main/resources/db/migration/V1__Init.sql [hunk 3]
```diff
@@ -0,0 +1,14 @@
+CREATE TABLE letter (
```
Comment (rkondratowicz) [inline src/main/resources/db/migration/V1__Init.sql]: Add some padding zeros in file name for easier sorting in the future

File: src/main/resources/db/migration/V1__Init.sql (hunk 2 above)
Comment (doncem) [reply src/main/resources/db/migration/V1__Init.sql]: how will querying work? is json type accommodated to perform efficiently?

Comment (rkondratowicz) [reply src/main/resources/db/migration/V1__Init.sql]: There are the `->` and `->>` operators

Comment (doncem) [reply src/main/resources/db/migration/V1__Init.sql]: the future 🐙

File: src/main/resources/db/migration/V001__Init.sql [hunk 4]
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE letter (
+  id UUID NOT NULL PRIMARY KEY,
+  service VARCHAR(256) NOT NULL,
+  created TIMESTAMP NOT NULL,
+  sent_to_print TIMESTAMP NULL,
+  printed TIMESTAMP NULL,
+  metadata JSON NULL
```
Comment (rkondratowicz) [inline src/main/resources/db/migration/V001__Init.sql]: it's called `additional_data` in the model class

File: src/main/resources/db/migration/V001__Init.sql [hunk 5]
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE letter (
+  id UUID NOT NULL PRIMARY KEY,
```
Comment (nitinprabhu) [inline src/main/resources/db/migration/V001__Init.sql]: are you` message_id ` is the `id` field here ?

File: src/main/resources/db/migration/V001__Init.sql [hunk 6]
```diff
@@ -0,0 +1,9 @@
+CREATE TABLE letter (
+  id BIGSERIAL NOT NULL PRIMARY KEY,
+  message_id UUID NOT NULL,
```
Comment (rkondratowicz) [inline src/main/resources/db/migration/V001__Init.sql]: message id is md5 hash. just varchar I guess...

Comment (doncem) [reply src/main/resources/db/migration/V001__Init.sql]: probably rename table now too

File: src/main/resources/db/migration/V001__Init.sql [hunk 7]
```diff
@@ -0,0 +1,9 @@
+CREATE TABLE message (
+  id UUID NOT NULL PRIMARY KEY,
+  message_id VARCHAR(256) NOT NULL,
+  service VARCHAR(256) NOT NULL,
+  created TIMESTAMP NOT NULL,
```
Comment (lgonczar) [inline src/main/resources/db/migration/V001__Init.sql]: What's the current convention of naming date columns? I like `created_at` or `created_date`, etc. (same with `printed`), but if we're already following the existing convention, then the change would make things inconsistent.

Comment (rkondratowicz) [reply src/main/resources/db/migration/V001__Init.sql]: +! for `xxx_at` (draft-store uses it)

**Pull request merged**